### PR TITLE
SKFunction Invoke Parameter Duplication

### PR DESCRIPTION
**Pull request by shawncal**
### Motivation and Context
Each of the SKFunction.Invoke calls have multiple ways of passing the same data. It's not clear that, if one way is specified, the other will be ignored.

https://github.com/microsoft/semantic-kernel/issues/727

### Description
This cleans things up such that you either pass a full `SKContext`, which has everything included (CancellationToken, Logger, input), or you pass each of these as separate args, and an SKContext is created for you.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [X] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [X] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

**Comment by shawncal** [comment]
@stephentoub PR addressing Issue https://github.com/microsoft/semantic-kernel/issues/727

File: dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs [hunk 1]
```diff
@@ -23,9 +23,9 @@ public sealed class ContextVariables : IEnumerable<KeyValuePair<string, string>>
     /// Constructor for context variables.
     /// </summary>
     /// <param name="content">Optional value for the main variable of the context.</param>
-    public ContextVariables(string content = "")
+    public ContextVariables(string? content = null)
     {
-        this._variables[MainKey] = content;
+        this._variables[MainKey] = content ?? string.Empty;
```
Comment (stephentoub) [inline dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs:28]: Is it important that the MainKey always has a value set for it?  If we didn't do this, _variables could be lazily-initialized only when variables were actually added.

File: dotnet/src/SemanticKernel/Planning/Plan.cs [hunk 2]
```diff
@@ -279,30 +280,35 @@ public async Task<Plan> InvokeNextStepAsync(SKContext context)
...
-        context ??= new SKContext(new ContextVariables(), null!, null, log ?? NullLogger.Instance, cancellationToken);
+        this.State.Update(input);
 
-        context.Variables.Update(input);
+        SKContext context = new(
+            this.State,
+            NullMemory.Instance,
+            null,
+            log ?? NullLogger.Instance,
+            cancellationToken);
```
Comment (stephentoub) [inline dotnet/src/SemanticKernel/Planning/Plan.cs:301]: The SKContext ctor isn't validating any of these arguments.  Should it be?

File: dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs [hunk 3]
```diff
@@ -155,44 +155,29 @@ public FunctionView Describe()
...
-        }
-
-        return this.IsSemantic
-            ? this.InvokeSemanticAsync(context, settings)
-            : this.InvokeNativeAsync(context);
+        SKContext context = new(
+            new ContextVariables(input),
+            NullMemory.Instance,
+            this._skillCollection,
+            log ?? NullLogger.Instance,
```
Comment (stephentoub) [inline dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs:177]: Do we expect end-user code to create SKContext instances?  I'm wondering if we should make it easier, e.g. allow the logger to be null and have the ctor pick a default logger if null was passed, or even just allow it to store null.

**Review by stephentoub — DISMISSED** [review]

File: dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs (hunk 3 above)
Comment (shawncal) [reply dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs:177]: yup, good call

File: dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs (hunk 1 above)
Comment (shawncal) [reply dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs:28]: done! update coming

**Review by dluc — COMMENTED** [review]
The changes to SKFunction will break scenarios where we pass the context only for the internal helpers, and want to pass the input as a string.

**Comment by shawncal** [comment]
> The changes to SKFunction will break scenarios where we pass the context only for the internal helpers, and want to pass the input as a string param.

Yeah I had the same consideration. This is another one in the theme of methods have multiple paths to provide the same arg, and it's unclear which will be used. If I set a context[Input] and then call a Invoke with a empty input (and a valid context), should the input be empty or what was there in context?

I think this makes the API nice and clean:
If you already have a context, go ahead and update it however you want, and pass that in.
If you don't, give us the pieces (input, vars, logger, cancel) -- most are optional -- and we'll build the context for you.
Make sense? I've got an updated PR to send out, and then I'll check in with you on the latest.

Flipping to draft until I have time to update.

**Comment by stephentoub** [comment]
> If you don't, give us the pieces (input, vars, logger, cancel) -- most are optional -- and we'll build the context for you.

It'd also be good for the SKContext.Skills collection to be non-nullable.  If no collection is supplied to the ctor, it can just initialize the Skills to an empty collection.  Then downstream consumption doesn't need to check whether the collection is null or not.

**Comment by shawncal** [comment]
> > If you don't, give us the pieces (input, vars, logger, cancel) -- most are optional -- and we'll build the context for you.
> 
> It'd also be good for the SKContext.Skills collection to be non-nullable. If no collection is supplied to the ctor, it can just initialize the Skills to an empty collection. Then downstream consumption doesn't need to check whether the collection is null or not.

Yup, good call! In fact, I've got something like this in my update... Because the IROSkillCollection interface is in Abstractions and the implementation is in SK.Core, I had to create a NullReadOnlySkillCollection in Abstractions. Just need to merge with that big skill collection change, and I'll get the update sent out.

File: dotnet/src/SemanticKernel.UnitTests/Planning/PlanTests.cs [hunk 4]
```diff
@@ -71,15 +71,6 @@ public async Task CanExecutePlanWithContextAsync()
         // Assert
         Assert.NotNull(result);
         Assert.Equal("Some input", result.Result);
-
```
Comment (lemillermicrosoft) [inline dotnet/src/SemanticKernel.UnitTests/Planning/PlanTests.cs:74]:
```suggestion

        plan = new Plan(goal);

        // Act
        context.Variables.Update("other input");
        result = await plan.InvokeAsync(context);

        // Assert
        Assert.NotNull(result);
        Assert.Equal("other input", result.Result);
```

File: dotnet/src/SemanticKernel.UnitTests/Planning/PlanTests.cs [hunk 5]
```diff
@@ -560,21 +551,20 @@ public async Task CanExecutePlanWithOneStepAndStateAsync()
...
         var mockFunction = new Mock<ISKFunction>();
-        mockFunction.Setup(x => x.InvokeAsync(It.IsAny<SKContext>(), null, null, default))
-            .Callback<SKContext, CompleteRequestSettings, ILogger, CancellationToken?>((c, s, l, ct) =>
+        mockFunction.Setup(x => x.InvokeAsync(It.IsAny<SKContext>(), null))
+            .Callback<SKContext, CompleteRequestSettings>((c, s) =>
                 returnContext.Variables.Update("Here is a poem about " + c.Variables.Input))
             .Returns(() => Task.FromResult(returnContext));
 
         var plan = new Plan(mockFunction.Object);
-        plan.State.Set("input", "Cleopatra");
```
Comment (lemillermicrosoft) [inline dotnet/src/SemanticKernel.UnitTests/Planning/PlanTests.cs:569]: I'd hope this one still works. Thoughts?

File: dotnet/src/SemanticKernel/Planning/Plan.cs (hunk 2 above)
Comment (shawncal) [reply dotnet/src/SemanticKernel/Planning/Plan.cs]: Updated with defaults for all values

**Pull request merged**